components/butterfly/src/server/mod.rs

```diff
@@ -1216,6 +1242,67 @@ impl<'a> Serialize for ServerProxy<'a> {
     }
 }
 
+// Note: this is a separate module solely to facilitate targeted
+// logging, e.g.
+//
+//     RUST_LOG=habitat_butterfly::server::election_trigger=trace
+mod election_trigger {
+    use habitat_common::FeatureFlag;
+    use std::{fs,
+              path::PathBuf};
+
+    /// If the HAB_FEAT_TRIGGER_ELECTION feature flag is enabled,
+    /// we'll look on disk to see if a sentinel file for the
+    /// triggering of an election of this service group is present. If
+    /// so, we should trigger that election.
+    ///
+    /// An election for service `foo.default` will be triggered if:
+    ///
+    /// - the supervisor is running with `HAB_FEAT_TRIGGER_ELECTION=1`
+    /// - the file `/hab/sup/default/data/trigger_foo.default_election` is present _and can also be
+    ///   successfully deleted by the Supervisor_.
+    ///
+    /// The latter is important to prevent repeated triggering of an
+    /// election. The user puts the file in place to signal their
+    /// intent, and the Supervisor deletes it as an acknowledgment of
+    /// that intent.
+    pub(super) fn maybe_trigger(service_group_name: &str,
+                                feature_flags: FeatureFlag,
+                                data_path: &Option<PathBuf>)
+                                -> bool {
+        if feature_flags.contains(FeatureFlag::TRIGGER_ELECTION) {
+            if let Some(ref path) = data_path {
+                let sentinel_file = path.join(format!("trigger_{}_election", service_group_name));
+                if sentinel_file.is_file() {
+                    if let Err(e) = fs::remove_file(&sentinel_file) {
+                        // If we trigger without being able to remove
+                        // the file, then we'll just constantly
+                        // trigger elections.
+                        warn!("HAB_FEAT_TRIGGER_ELECTION: Could not delete sentinel file {:?}; \
+                               NOT triggering election! {:?}",
+                              sentinel_file, e);
+                        false
+                    } else {
+                        info!("HAB_FEAT_TRIGGER_ELECTION: Manually triggering election for {}",
+                              service_group_name);
+                        true
+                    }
+                } else {
+                    trace!("HAB_FEAT_TRIGGER_ELECTION: No trigger file found for {}",
+                           service_group_name);
+                    false
+                }
+            } else {
+                trace!("HAB_FEAT_TRIGGER_ELECTION: In a test");
+                false
+            }
+        } else {
+            trace!("HAB_FEAT_TRIGGER_ELECTION: Feature not enabled");
+            false
+        }
+    }
```

Comment:
My feelings on this subject aren't super strong, but it occurs to me reading through this that you could likely split out at least a few of these if statements into early returns and avoid having things so deeply nested.